### PR TITLE
Feature: lazily generate mindsphere datalake presign url 

### PR DESCRIPTION
**Pull request by lucian-torje-siemens**
## What this PR changes/adds

Enable creating download presign urls before using them and enable uploading to cloud (already generated) upload urls.

## Why it does that

We need it so that we can enable our customers to provide their own cloud solutions as input/output for data 

## Further notes

New extension added to provide support presigned url download (already enabled using HttpData) and upload (missing from HttpData)

## Linked Issue(s)

Closes # <-- _insert Issue number if one exists_

## Checklist

- [ ] added appropriate tests?
- [ ] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [ ] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeClientImpl.java [hunk 1]
```diff
@@ -15,45 +15,161 @@
...
 public class DataLakeClientImpl implements DataLakeClient {
-    public DataLakeClientImpl(URI dataLakeBaseUrl, OkHttpClient client, ObjectMapper objectMapper) {
+    public DataLakeClientImpl(URL dataLakeBaseUrl, OkHttpClient client, ObjectMapper objectMapper) {
         this.dataLakeBaseUrl = dataLakeBaseUrl;
         this.client = client;
         this.objectMapper = objectMapper;
     }
 
-    private final URI dataLakeBaseUrl;
+    public DataLakeClientImpl(URL dataLakeBaseUrl) {
```
Comment (alex-ghiran-siemens) [inline extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeClientImpl.java]: delete this constructor as it is unused

File: extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeClientImpl.java [hunk 2]
```diff
@@ -15,45 +15,161 @@
...
+/**
+ * It has the responsibility to:
+ * - create upload presign url (used to upload files to)
+ * - create download presign url (used to get the file content from)
+ *
+ * It needs a
+ */
 public class DataLakeClientImpl implements DataLakeClient {
-    public DataLakeClientImpl(URI dataLakeBaseUrl, OkHttpClient client, ObjectMapper objectMapper) {
+    public DataLakeClientImpl(URL dataLakeBaseUrl, OkHttpClient client, ObjectMapper objectMapper) {
```
Comment (alex-ghiran-siemens) [inline extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeClientImpl.java]: Delete this constructor after moving the client and objectMapper assignment to the only used constructor

File: extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeClientImpl.java [hunk 3]
```diff
@@ -15,45 +15,161 @@
...
+
+            final String extractedTenant = extractTenantFromPath(path)
+                    .orElseThrow(() -> new DataLakeException("Could not identify tenant for path " + datalakePath));
+
+            final String accessToken = getAccessToken(extractedTenant)
+                    .orElseThrow(() -> new DataLakeException("Getting token failed"));
+
+            final String payloadString = objectMapper.writeValueAsString(new DatalakePaths()
+                    .setPaths(Collections.singletonList(new DatalakePath().setPath(extractFilenameFromPath(datalakePath)))));
+            final RequestBody requestPayload = RequestBody.create(payloadString, MediaType.parse("application/json"));
```
Comment (alex-ghiran-siemens) [inline extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeClientImpl.java]: Use a constant for application/json

File: extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeClientImpl.java [hunk 4]
```diff
@@ -15,45 +15,161 @@
...
+            final Request request = new Request.Builder()
+                    .method("POST", requestPayload)
```
Comment (alex-ghiran-siemens) [inline extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeClientImpl.java]: use a constant for POST

File: extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeClientImpl.java [hunk 5]
```diff
@@ -72,6 +188,49 @@ public URL getUploadUrl(String path) throws DataLakeException {
...
+        TechnicalUserTokenRequestDto technicalUserTokenRequestDto = new TechnicalUserTokenRequestDto(oauthClientDetails.getTenant(),
+                tenant, oauthClientDetails.getClientAppName(), oauthClientDetails.getClientAppVersion());
+
+        final String requestPayload = objectMapper.writeValueAsString(technicalUserTokenRequestDto);
+        final RequestBody requestBody = RequestBody.create(requestPayload, MediaType.parse("application/json"));
+
+
+        final Request tokenRequest = new Request.Builder().url(oauthClientDetails.getAccessTokenUrl())
+                .method("POST", requestBody)
+                .header(X_SPACE_AUTH_KEY, String.format("Bearer %s", oauthClientDetails.getBase64Credentials()))
```
Comment (alex-ghiran-siemens) [inline extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeClientImpl.java]: Replace "Bearer " with a constant

File: extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeExtension.java [hunk 6]
```diff
@@ -37,40 +47,109 @@ public class DataLakeExtension implements ServiceExtension {
...
+            final String tokenmanagementClientId = context.getSetting(TOKENMANAGEMENT_CLIENT_ID, "");
+            final String tokenmanagementClientSecret = context.getSetting(TOKENMANAGEMENT_CLIENT_SECRET, "");
+            final String tokenmanagementClientAppName = context.getSetting(TOKENMANAGEMENT_CLIENT_APP_NAME, "");
+            final String tokenmanagementClientAppVersion = context.getSetting(TOKENMANAGEMENT_CLIENT_APP_VERSION, "");
+            final String tokenmanagementAddress = context.getSetting(TOKENMANAGEMENT_ADDRESS, "");
+            final String dataLakeAddress = context.getSetting(DATALAKE_BASE_URL, "");
+
+            final String applicationTenant = context.getSetting(APPLICATION_TENANT, "presdev");
+
+            monitor.debug("datalake.oauth.clientId=" + tokenmanagementClientId.substring(0, 5));
```
Comment (alex-ghiran-siemens) [inline extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeExtension.java:92]: Why are you logging only the first 5 characters?

File: extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/provision/SourceUrlResourceDefinitionGenerator.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    @Override
+    public @Nullable ResourceDefinition generate(DataRequest dataRequest, DataAddress assetAddress, Policy policy) {
+        if (dataRequest.getDestinationType() == null) {
+            monitor.debug("There is no destination type");
+            return null;
+        }
+
+        final String dataDestinationType = dataRequest.getDataDestination().getType();
+        if (!HttpDataAddressSchema.TYPE.equals(dataDestinationType) &&
+                !MindsphereSchema.TYPE.equals(dataDestinationType)) {
```
Comment (alex-ghiran-siemens) [inline extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/provision/SourceUrlResourceDefinitionGenerator.java]: Maybe "!List.of(HttpAddressSchema.TYPE, MindsphereSchema.TYPE).contains(dataDestinationType)" as a condition is more readable

File: samples/other/http-transfer-provisioning/provider/src/main/java/com/siemens/mindsphere/PrepareContractExtension.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+    @Override
+    public String name() {
+        return "Default dataplane httpdata transfer with provision on consumer side";
+    }
+
+    private Policy createPolicy() {
+
+        var usePermission = Permission.Builder.newInstance()
+                .action(Action.Builder.newInstance().type("USE").build())
```
Comment (alex-ghiran-siemens) [inline samples/other/http-transfer-provisioning/provider/src/main/java/com/siemens/mindsphere/PrepareContractExtension.java]: This "USE" value seems to be used in the same manner in multiple classes. Consider creating a constant in an appropriate class and using a reference to that constant everywhere

**Review by alex-ghiran-siemens — APPROVED** [review]

File: extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeExtension.java [hunk 9]
```diff
@@ -37,40 +47,109 @@ public class DataLakeExtension implements ServiceExtension {
...
-        // register Data Lake Reader
-        dataOperatorRegistry.registerReader(dataLakeReader);
+    private void registerProvisionDefinitionsGenerators(ServiceExtensionContext context, Monitor monitor) {
+        // register the datalake resource definition generator
+        manifestGenerator.registerGenerator(new SourceUrlResourceDefinitionGenerator(monitor, context));
 
-        // register provisioner
-        @SuppressWarnings("unchecked") var retryPolicy = (RetryPolicy<Object>) context.getService(RetryPolicy.class);
+        // register the datalake resource definition generator
+        manifestGenerator.registerGenerator(new SourceUrlResourceDefinitionGenerator(monitor, context));
```
Comment (lucian-torje-siemens) [inline extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeExtension.java]: here it should be DestinationUrlDefinitionGenerator

File: extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeExtension.java (hunk 6 above)
Comment (lucian-torje-siemens) [reply extensions/mindsphere/mindsphere-http/src/main/java/com/siemens/mindsphere/datalake/edc/http/DataLakeExtension.java:92]: security issues

**Pull request merged**